### ISSUE TITLE
[FEATURE] Live migration progress awareness

### ISSUE THREAD

**Issue by Martin-Weiss** [harvester/harvester#4352]
**Is your enhancement related to a problem? Please describe.**
We have the problem over and over again, that the migration does get stuck and does not complete. In this situation the UI does not help at all to see where the problem might be.. what status the migration is in and how long it might take etc.

**Describe the solution you'd like**
In the UI when doing a migration indicate the steps that are done during a migration (create new vm, copy memory, show pogress of memory copy, ....)

**Describe alternatives you've considered**
did not find an alternative

**Additional context**
mayhe these metrics can help as well?

https://github.com/kubevirt/kubevirt/pull/7474
https://github.com/kubevirt/kubevirt/pull/7946

**Comment by FrankYang0529** [comment]
After some investigation, we can't get migration progress from [VirtualMachineInstance.Status.MigrationState](https://github.com/kubevirt/kubevirt/blob/e4c7d4e53269d2c416a4daae761b262b668529ce/staging/src/kubevirt.io/api/core/v1/types.go#L637-L684). If we look into code, we can find [virt-launcher gets migration status](https://github.com/kubevirt/kubevirt/blob/e4c7d4e53269d2c416a4daae761b262b668529ce/pkg/virt-launcher/virtwrap/live-migration-source.go#L686-L699) from [Domain](https://gitlab.com/libvirt/libvirt-go-module/-/blob/master/domain.go?ref_type=heads#L3209-3243). If we can find a way to get connection through virt-api -> virt-handler -> virt-launcher, we can use `MemTotal` and `MemProcessed` to get the progress.

**Comment by ibrokethecloud** [comment]
we should also leverage the conditions on the `vmim` object to indicate the various phases of the virtual machine migration.

For example in certain cases due to cluster specific defaults of 5 parallel migrations and 2 migrations per target node, a number of migrations may be queued.

We should ensure that we indicate that the migration is queued, so users know that the migration is scheduled but has not yet started.

**Comment by innobead** [comment]
The backend effort was finished, and the current status is pending for UI improvement, https://github.com/harvester/harvester/issues/7339.  cc @harvester/ui

**Comment by TachunLin** [comment]
### Test Scenarios

1. Migration success scenario
    - Check migration details
    - Check the Grafana dashboard 
  
2. Migration failed scenario
    - Check migration details
    - Check the Grafana dashboard 
  
3. Abort Migration scenario
    - Check migration details
    - Check the Grafana dashboard 

4. Migration not supported for the node scheduling rule
    - Check migration details
    - Check the Grafana dashboard 

5. Upgrade path scenario 
    - https://github.com/harvester/harvester/pull/7551
    - Check after upgrade Harvester can well display vm migration data in the Grafana page

**Comment by TachunLin** [comment]
* Since on the current v1.6-head three nodes cluster, the vm migration would stuck in `Migrating` until timeout.

  <img width="1599" height="786" alt="Image" src="https://github.com/user-attachments/assets/9ee8029b-78ab-4b68-aad7-22ecf98252b8" />
  
* This issue verification would been blocked by issue #8617. 
* Thus move back to ready for testing.

**Comment by TachunLin** [comment]
After build another three nodes Harvester on local bare-metal machines with latest v1.6-head.
Retest there can well migrate vm without stuck in Migrating.
Thus move this issue to testing and continue testing on that environment.

**Comment by TachunLin** [comment]
After discussed with team. 
Currently the expected behavior of each migration scenarios would be:


1. Migration `success` (enabled monitoring addon)
    - Display phase, source/target node and started/ended time
    - Display grafana monitoring chart
    - Display events at the bottom (Diminish after successful migration)   
2. Migration `failed` (enabled monitoring addon)
    - Display phase, source/target node and started/ended time
    - Display grafana monitoring chart
    - Display events at the bottom (Diminish after successful migration)   

3. Migration `success` (No monitoring addon)
    - Display phase, source/target node and started/ended time
    - Display events at the bottom (Diminish after successful migration)   
4. Migration `failed` (No monitoring addon)
    - Display phase, source/target node and started/ended time
    - Display events at the bottom (Diminish after successful migration)   

5. `Abort` Migration

    - Won't display or update any content in the Migration details page

6. Migration `not supported` for the node scheduling rule

    - Won't display or update any content in the Migration details page

**Comment by TachunLin** [comment]
From the video clip of the migration `success` scenario:

1. It can display the  phase, source/target node and started/ended time 
2. But the target node information is not correct ( should be node1 but display node3 instead)
3. The migration event would automatically disappear after finished migration 

  https://github.com/user-attachments/assets/6d70774f-56d4-463c-ae6e-efaca93b2bc6
  
  https://github.com/user-attachments/assets/d8dda73b-6292-4698-926b-02a93f8c82ab
  
  <img width="1696" height="1035" alt="Image" src="https://github.com/user-attachments/assets/d8d0398c-fddc-490d-afdf-75192b02cd17" />

**Comment by TachunLin** [comment]
On one of the test environments that can reproduce failure vm migration.

From the video clip, when vm migration failed

1. There is no any content recorded in the vm migration details page

  https://github.com/user-attachments/assets/e1e402e4-5152-4afb-904c-8f3f4d11a75c
  
  https://github.com/user-attachments/assets/bed73255-0028-4d87-b3ea-4db7a206e9de

  <img width="720" height="358" alt="Image" src="https://github.com/user-attachments/assets/3b399a95-0ca4-4b05-9bf2-aa8880674032" />

**Comment by TachunLin** [comment]
From another test environment that can reproduce failure migration.

On this environment, when migration failed

1. It can display the phase, source/target node and started/ended time
2. But the target node information is not correct ( should be node1 but display node3 instead)
3. The migration event can keep in the details after then automatically cleanup when time is up.

  https://github.com/user-attachments/assets/a11efe4f-0c8e-4669-8630-ca13a50060ce
  

  <img width="1916" height="1076" alt="Image" src="https://github.com/user-attachments/assets/b4f9c476-65ff-44e4-b64d-6ed87f194af7" />

**Comment by TachunLin** [comment]
According to these test result, the mechanism to handle and keep migration failure is not stable.
Thus move to ready for testing for further enhancement.

**Comment by TachunLin** [comment]
After the PR fix: https://github.com/harvester/harvester-ui-extension/pull/390
Now we can correctly get the migrate state and events in the migration details page.

  https://github.com/user-attachments/assets/1e952bf0-f1f4-4393-a69a-6e5b1675f1f7
  
  <img width="1689" height="1044" alt="Image" src="https://github.com/user-attachments/assets/574b64b4-c3aa-4d76-9a54-5208b948cd46" />

**Comment by TachunLin** [comment]
While we encounter another issue during testing.

* We found the vm did not actually migrate to the target node we choose in the list.

With a clean 3 nodes `v1.6.0-rc1` Harvester cluster (disable rancher-monitoring addon): 

In the video, we tried to migrate to **node1** but it actually migrate to **node3** 

* support bundle: [supportbundle_78b9afdf-52b8-4905-8f08-3ddb97924277_2025-07-14T06-17-59Z.zip](https://github.com/user-attachments/files/21212144/supportbundle_78b9afdf-52b8-4905-8f08-3ddb97924277_2025-07-14T06-17-59Z.zip)


  https://github.com/user-attachments/assets/b05226a3-3b90-44a7-8889-2a744d74fe26
  

With the same 3 nodes `v1.6.0-rc1` Harvester cluster (enabled rancher-monitoring addon): 

In the video, we tried to migrate to **node1** but it still migrate to **node3** 

* support bundle: [supportbundle_78b9afdf-52b8-4905-8f08-3ddb97924277_2025-07-14T06-29-13Z.zip](https://github.com/user-attachments/files/21212167/supportbundle_78b9afdf-52b8-4905-8f08-3ddb97924277_2025-07-14T06-29-13Z.zip)

  https://github.com/user-attachments/assets/44ff932c-16e4-4dac-90bd-ce9a9af5d369
  

On another 3 nodes kvm environment also encounter this issue

We tried to migrate to `harvester-node-1` but actually migrate to `harvester-node-2`


  https://github.com/user-attachments/assets/57161e37-2149-40e1-b2bc-84411053fb55

**Comment by TachunLin** [comment]
We need to wait for the migrate to incorrect target issue to be fixed then can continue the rest of the testing.
*   https://github.com/harvester/harvester/issues/8670

Thus move this issue to ready for testing.

**Comment by TachunLin** [comment]
Verifying on `v.1.6.0-rc3` with Rancher `v2.12.0-rc2`.

Blocked by issue: https://github.com/harvester/harvester/issues/8750.
Move back to ready for testing.

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Success migration can display correct target, time, chart and events $~~$</summary>

1. When migration success, on the migration details, it can display the correct source and target node
2. Display the migration chart 
3. Display the migration events

  <img width="1529" height="789" alt="Image" src="https://github.com/user-attachments/assets/216ed883-f214-4786-abaa-48826aff751d" />
  
  <img width="1528" height="790" alt="Image" src="https://github.com/user-attachments/assets/351632c7-ae9c-49f5-8cdf-a5ee7afaa324" />
  
  
  https://github.com/user-attachments/assets/e6cada15-6531-4298-af94-34d815af34e9

  https://github.com/user-attachments/assets/c938507b-4681-497b-a847-e9dfd4ff6406
  
  
  
 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Failed migration can display correct target, time, chart and events $~~$</summary>

1. When migration success, on the migration details, it can display the correct source and target node
2. Display the migration chart 
3. Display the migration events

  <img width="1907" height="1085" alt="Image" src="https://github.com/user-attachments/assets/38c32801-2420-4da7-ab21-06dc3e30c144" />
  
  
  <img width="1526" height="780" alt="Image" src="https://github.com/user-attachments/assets/6d60fbe6-83b7-4f51-8869-b321fd8c0b2c" />
  
  
  <img width="1599" height="784" alt="Image" src="https://github.com/user-attachments/assets/7160262f-e7a1-4477-bcf1-3ee0a144fe34" />
  
  
  https://github.com/user-attachments/assets/9e1620c5-24f1-487e-9d09-a763f8a65c8d

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Abort migration won't show phase, would keep the original node and time events $~~$</summary>

1. When we abort migration on the migration details, it will clean up the phase and keep other information
2. Display the migration chart 
3. Display the migration events

  <img width="1918" height="1084" alt="Image" src="https://github.com/user-attachments/assets/00eaf1fd-1177-4f98-903f-5d07d4c69abd" />
  
  https://github.com/user-attachments/assets/ba53a6b6-deb7-4d4b-84e8-3ae3230eece1

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Can display migration chart from vm migration details -> Grafana icon $~~$</summary>

  <img width="1916" height="799" alt="Image" src="https://github.com/user-attachments/assets/349118a5-5d1f-4ed1-a37e-a4bc1d566bcc" />

 </details>

  <details><summary>$\color{green}{\textsf{PASS}}$ Can display migration chart from Rancher-monitoring > Grafana > click Grafana icon > Dashboards > Harvester VM Migration Details $~~$</summary>
  
  <img width="1920" height="1081" alt="Image" src="https://github.com/user-attachments/assets/a298e7c9-e833-4412-baa9-9e99f072ffaa" />
  
 </details>

<details><summary>$\color{red}{\textsf{Blocked}}$ Test after Harvester upgrade to v1.6.0-rc $~~$</summary>

  * Migration scenarios blocked by https://github.com/harvester/harvester/issues/8750

 </details>


### Test Information
* Test Environment: Three nodes bare-metal machines
* Harvester version: v1.6.0-rc3
* Rancher version: v2.12.0-rc2
* Harvester UI extension: v1.6.0-rc3

### Verify Steps
<details><summary> Check migration details and migration chart</summary>

  1. Prepare a three nodes Harvester cluster
  2. Enable the `rancher-monitoring` addon
  3. Create an ubuntu image
  4. Create the vm named `vm1`
  5. ssh to vm1 
  6. Execute the stress tool

  ```
  sudo apt install stress
  stress --vm 2 --vm-bytes 100M --timeout 86400s
  ```

  7. Migrate `vm1` to other node
  8. Need to test the migration succeed and failed scenarios
  9. Check the vm details -> Migration 
  10. Check the source and target node 
  11. Check the start and end time 
  12. Check the migration chart 
  13. Check the migration events
  14. Click the grafana icon, check can direct to the Grafana Harvester VM Migration Details
  15. Check Add-ons > rancher-monitoring > Grafana > click Grafana icon > Dashboards > Harvester VM Migration Details

</details>

**Comment by TachunLin** [comment]
## Retest on v1.6.0-rc4 for the upgrade and fresh install Harvester


### After upgrade from v1.5.1 to v1.6.0-rc4

### Result
<details><summary>$\color{green}{\textsf{PASS}}$ Migrate existing vm without rancher-monitoring addon enabled $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

  <img width="1600" height="790" alt="Image" src="https://github.com/user-attachments/assets/6d5311df-1d26-4180-b7f8-eaeb26217212" />
  
  https://github.com/user-attachments/assets/d07e0380-2bd8-4cef-8e12-ed228bd64c64

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Migrate new vm without rancher-monitoring addon enabled $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

  <img width="1598" height="788" alt="Image" src="https://github.com/user-attachments/assets/6b7a5f81-c210-4d63-a63d-82f1fa790dc6" />
  
  https://github.com/user-attachments/assets/cce0f472-4026-49df-b3b2-ee5798127c24

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Migrate new vm with rancher-monitoring addon enabled $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

  <img width="1598" height="788" alt="Image" src="https://github.com/user-attachments/assets/b715089b-3bb5-476e-949b-f6d24a8c6ded" />
  
  <img width="1600" height="785" alt="Image" src="https://github.com/user-attachments/assets/87d78616-9ccd-4cdc-80e0-1ff1004f0a30" />
  
  https://github.com/user-attachments/assets/30494b4a-0392-4c6d-9d3d-b61712f69f8f

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Migrate existing vm with rancher-monitoring addon enabled  $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

  <img width="1598" height="793" alt="Image" src="https://github.com/user-attachments/assets/f22f864f-b695-4179-98a7-c5f3c954fba6" />
  
  <img width="1600" height="792" alt="Image" src="https://github.com/user-attachments/assets/1ace2164-b8fb-4bc9-86f9-6d427addb7ff" />
  
  https://github.com/user-attachments/assets/167dc92e-8bcd-42e7-b7ee-ee5b2e635d97

 </details>

--- 

### Brand new installation of v1.6.0-rc4

<details><summary>$\color{green}{\textsf{PASS}}$ Migrate existing vm with rancher-monitoring addon enabled  $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

  <img width="1917" height="1081" alt="Image" src="https://github.com/user-attachments/assets/872f3679-8feb-4c93-84c0-426d7ed6c852" />
  
  https://github.com/user-attachments/assets/6ff92d49-e792-48ba-9123-e5c6cf7980e8


  https://github.com/user-attachments/assets/d3301175-9bee-4630-a0f8-cb3ad3872121
  
  <img width="1917" height="1082" alt="Image" src="https://github.com/user-attachments/assets/9cbd6332-adcb-45a1-9759-965121823f39" />

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Migrate new vm with rancher-monitoring addon enabled  $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

<img width="1917" height="1085" alt="Image" src="https://github.com/user-attachments/assets/3fad8380-6fa7-404b-9fc0-59346d565a06" />

<img width="1920" height="1085" alt="Image" src="https://github.com/user-attachments/assets/96cab096-4d39-45f6-845a-4f2c26a93e93" />

https://github.com/user-attachments/assets/ca1979bc-172c-4d4d-8088-2d2d962ec180

 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Migrate new vm again to other node after 30 minutes $~~$</summary>

  1. Can correctly migrate vm to the target node
  2. Can display correct source and target node information
  3. Can display correct start and end time

    <img width="1914" height="1080" alt="Image" src="https://github.com/user-attachments/assets/968a1443-fdb9-422d-afd6-4538d35bfbf0" />
    
    <img width="1918" height="1087" alt="Image" src="https://github.com/user-attachments/assets/f2863871-f9a7-4a75-b443-24bea2a6e54d" />
  
  https://github.com/user-attachments/assets/31c22dca-706d-4a72-99c8-af4e0ac45dfa

 </details>

### Test Information
* Test Environment: Three nodes bare-metal and kvm machines
* Harvester version: v1.6.0-rc4

### Verify Steps
<details><summary> Upgrade path </summary>

  1. Prepare a three nodes v1.5.1 Harvester
  2. Upgrade to v1.6.0-rc4, (Apply workaround in [#8773](https://github.com/harvester/harvester/issues/8773#issue-3273867202))
  3. Create a vm 
  4. Migrate vm to other node
  5. Check the migration details
  6. Enable rancher-monitoring addon
  7. Create another vm
  8. Migrate vm to other node
  9. Check the migration details

</details>

**Comment by TachunLin** [comment]
This time when retesting the upgrade path on v1.6.0-rc4.

I tried to use the minimum requirement setup to verify this feature to ensure migration would succeed under different scenarios. 

Thus I remove the following steps and **only** enable the rancher-monitoring addon

1. Did not have Rancher integrated with Harvester
2. Did not set the restrict migration bandwidth to 2Mi 
    ```
    kubectl patch kubevirt kubevirt -n harvester-system --type merge -p '{"spec":{"configuration":{"migrations":{"bandwidthPerMigration":"2Mi"}}}}'
    ```

3. Did not run the stress test tool on the vm
    ```
    sudo apt install stress
    stress --vm 2 --vm-bytes 100M --timeout 86400s
    ```

Since we can pass the test scenarios, I think we are good to close this issue. 
For the vm migration under complex environment and factor, we can continue to track in issue  https://github.com/harvester/harvester/issues/8617
